openapi tutorial layout

thimy opened the pull request:
_This PR depends on #277_

* Added parsing of yml file to generate guide sidebar the same way we do for help sidebar
* Created the pages necessary to display the newly created openapi guides.

I left Phil's commits so we can check how it renders but what really counts is the last commit, I'll remove them once this gets approved.

There's also a commit that fixes links in markdown that will go to #256.
@ChristopheDujarric you can start reviewing Phil's posts from here but comments will have to go to #256.

ChristopheDujarric commented:
Not sure how far this relates to this PR, but there is an issue with wide tables. Example: https://deploy-preview-279--bump-content-hub.netlify.app/guides/openapi/specification/v3.1/understanding-structure/parameter-serialization/#query-parameters

![image](https://github.com/user-attachments/assets/1a58e1ae-3c61-4d5c-aaa6-62f5996250e1)

Should we consider larger content zones?

ChristopheDujarric commented:
It isn't clear in mind mind anymore: did we not want to keep the "next article"/"previous article at the end of pages?

This, or have the scroll only on the content area, not on the side bar, so that the index is always visible.
![image](https://github.com/user-attachments/assets/8ef7876a-6ff1-4e73-81c3-6876e64db9ab)

ChristopheDujarric commented:
Apart from those two comments, this looks gorgeous, thanks @thimy and @grossyoan !

ChristopheDujarric commented on src/_data/sidebars/guides/openapi/v3-1.yml:
```diff
@@ -0,0 +1,75 @@
+    items:
+      - label: Basic Structure
+        link: /understanding-structure/basic-structure/
+      - label: History and Evolution of OpenAPI
```
```suggestion
      - label: Defining API Servers
```

ChristopheDujarric commented on src/_data/sidebars/guides/openapi/v3-1.yml:
```diff
@@ -0,0 +1,75 @@
+        link: /understanding-structure/basic-structure/
+      - label: History and Evolution of OpenAPI
+        link: /understanding-structure/api-servers/
+      - label: Benefits of Using OpenAPI
```
```suggestion
      - label: Paths and Operations
```

thimy commented:
Sorry, I didn't want to show entrypoints to the guide while they were not approved but that was a bit stupid since you can't review the layout. xd
The last commit will need to be removed before merging.

thimy commented:
I have added the prev/next navigation and added css classes to add on a per table basis to fix the overflowing content.

![image](https://github.com/user-attachments/assets/ca7a90a7-e2da-44d8-aff8-d9aadba3633b)

grossyoan commented:
Perfect!

Could you rename "OpenAPI" to "OpenAPI Specification" in the navbar? So we're good on that part too.

Thanks a lot for all this good work Thimy!

thimy commented:
Done ✅

thimy commented:
I removed Phil's commits to avoid noise, may I have a technical review please? 🙏

fbraure commented on src/_components/guides/in_depth/post_info.erb:
```diff
@@ -0,0 +1,16 @@
+<indepth-post-info>
+  <div class="author">
+    By
+    <% @authors.each do |author| %>
```
If there is several authors, don't you need separator between them?

fbraure commented on src/_components/guides/in_depth/post_info.rb:
```diff
@@ -0,0 +1,22 @@
+class Guides::InDepth::PostInfo < Bridgetown::Component
+  def initialize(authors:, last_update:)
+    @authors = authors.class == String ? [authors] : authors
```
```suggestion
    @authors = authors.is_a?(String) ? [authors] : authors
```

fbraure replied:
This switch is not a good practice. We often don't have choice. 
May be you can write: 
```suggestion
    @authors = Array(authors)
```
<img width="358" alt="image" src="https://github.com/user-attachments/assets/8119be2a-f17c-46d9-be77-925698c92f32">

fbraure commented on src/_components/guides/in_depth/post_info.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def is_guest?(author)
+    author_data(author).class == String
```
```suggestion
    author_data(author).is_a?(String)
```

fbraure commented on plugins/builders/sidebar.rb:
```diff
@@ -1,23 +1,29 @@
   end
 
-  def build_resources(list)
+  def build_resources(list, collection, root)
```
`list` and `collection` are very generic names, difficult to understand and distinguish. Do you have more explicit names in mind?

fbraure replied:
`Collection` seems to be a Bridgetown word. We can just explicit `list`

fbraure commented on src/_layouts/indepth_guide.erb:
```diff
@@ -0,0 +1,25 @@
+  <% end %>
+<% end %>
+<main class="guide-content wrapper" data-wrapper-size="small">
+  <button data-button-style="secondary" data-button-size="small" data-toggle-target="button" data-action="toggle#click">
```
You don't need to specify the controller for the `data-action`?

thimy replied:
The controller is `toggle`, I vonluntarily kept a very generic name because it just toggles any stuff given a button and a target.

fbraure commented on src/_layouts/indepth_guide.erb:
```diff
@@ -0,0 +1,25 @@
+  <button data-button-style="secondary" data-button-size="small" data-toggle-target="button" data-action="toggle#click">
+    Show menu <svg width="11" height="12" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="m10 6 .543.517a.75.75 0 0 0 0-1.034L10 6ZM0 6.75h10v-1.5H0v1.5Zm10.543-1.267-4.762-5-1.086 1.034 4.762 5 1.086-1.034Zm-1.086 0-4.762 5 1.086 1.034 4.762-5-1.086-1.034Z" fill="#025FD7"/></svg>
+  </button>
+  <h1><%= data.title %></h1>
```
You are gona have 2 <h1> in the page?

thimy replied:
I'm not sure I understand what you mean, we always skip adding `#` in markdown content and we put the main title in the frnotmatter in case we need to retrieve it as data.

fbraure replied:
As discussed, if we have some tags that reset the h* hierarchy, like section define in the Sidebar component, that's not a subject

fbraure reviewed (approved):
That ok for me, thank you Thimy :)
